packages/devtools_shared/lib/src/test/integration_test_runner.dart

```diff
@@ -0,0 +1,335 @@
+                !line.contains(_beginExceptionMarker)) {
+              stdOutWriteInProgress = false;
+              exceptionBuffer.writeln();
+            }
+          }
+        },
+        onStderr: (line) {
+          if (line.contains(_errorMarker) ||
+              line.contains(_unhandledExceptionMarker)) {
+            stdErrWriteInProgress = true;
+          }
+          if (stdErrWriteInProgress) {
+            exceptionBuffer.writeln(line);
+          }
+        },
+      );
+
+      bool testTimedOut = false;
+      final timeout = Future.delayed(const Duration(minutes: 8)).then((_) {
+        testTimedOut = true;
+      });
+
+      await Future.any([
+        process.exitCode,
+        timeout,
+      ]);
+
+      debugLog('attempting to kill the flutter drive process');
+      process.kill();
+      debugLog('flutter drive process has exited');
+
+      // Ignore exception handling and retries if the tests passed. This is to
+      // avoid bugs with the test runner where the test can fail after the test
+      // has passed. See https://github.com/flutter/flutter/issues/129041.
+      if (!testsPassed) {
+        if (testTimedOut) {
+          if (attemptNumber >= _maxRetriesOnTimeout) {
+            throw Exception(
+              'Integration test timed out on try #$attemptNumber: $testTarget',
+            );
+          } else {
+            debugLog(
+              'Integration test timed out on try #$attemptNumber. Retrying '
+              '$testTarget now.',
+            );
+            await runTest(attemptNumber: ++attemptNumber);
+          }
+        }
+
+        if (exceptionBuffer.isNotEmpty) {
+          throw Exception(exceptionBuffer.toString());
+        }
+      }
+    }
+
+    await runTest(attemptNumber: 0);
+  }
+}
+
+class _IntegrationTestResult {
+  _IntegrationTestResult._(this.result, this.methodName, this.details);
+
+  factory _IntegrationTestResult.parse(Map<String, Object?> json) {
+    final result = json[resultKey] == 'true';
+    final failureDetails =
+        (json[failureDetailsKey] as List<Object?>).cast<String>().firstOrNull ??
+            '{}';
+    final failureDetailsMap =
+        jsonDecode(failureDetails) as Map<String, Object?>;
+    final methodName = failureDetailsMap[methodNameKey] as String?;
+    final details = failureDetailsMap[detailsKey] as String?;
+    return _IntegrationTestResult._(result, methodName, details);
+  }
+
+  static const testResultPrefix = 'result {"result":';
+  static const resultKey = 'result';
+  static const failureDetailsKey = 'failureDetails';
+  static const methodNameKey = 'methodName';
+  static const detailsKey = 'details';
+
+  final bool result;
+  final String? methodName;
+  final String? details;
+
+  @override
+  String toString() {
+    if (result) {
+      return 'Test passed';
+    }
+    return 'Test \'$methodName\' failed: $details.';
+  }
+}
+
+class IntegrationTestRunnerArgs {
+  IntegrationTestRunnerArgs(
```

Comment:
base class created from the original TestRunnerArgs class in run_tests.dart. Stripped away any args that were specific to devtools app for this base class